### DISCUSSION TITLE
contentType SNS attribute required to deserialize JSON with @NotificationMessage: bug or not?

### DISCUSSION THREAD

**Discussion by hertzsprung** [awspring/spring-cloud-aws#520] [Q&A] [answered]
Using Spring Cloud AWS 2.4.2, publishing a JSON message to an SNS topic leads to this `SqsListener` receiving a `null` payload:

```java
@SqsListener(value = "myQueue")
public void process(@NotificationMessage MyJsonPayload payload) {
    // ...
}
```
After some [source](https://github.com/awspring/spring-cloud-aws/blob/2.4.x/spring-cloud-aws-messaging/src/main/java/io/awspring/cloud/messaging/config/QueueMessageHandlerFactory.java#L201) [diving](https://github.com/awspring/spring-cloud-aws/blob/2.4.x/spring-cloud-aws-messaging/src/main/java/io/awspring/cloud/messaging/support/converter/NotificationRequestConverter.java#L60) I realised the SNS message must specify an attribute `contentType: application/json` to be recognised by the `MappingJackson2MessageConverter`.

Is this intentional (undocumented?) behaviour or a bug?

I worked around the problem by declaring my own `MappingJackson2MessageConverter` bean:
```java
@Bean
MappingJackson2MessageConverter converter() {
    return new MappingJackson2MessageConverter(new MimeType[]{});
}
```

**Comment by maciejwalkowiak** [comment, marked as answer]
@hertzsprung which version of Spring Cloud AWS are you using?

Edit:

Enforcing content type match is already fixed in 3.0 M2 (https://github.com/awspring/spring-cloud-aws/issues/213). You can follow the discussion there for more details.

**Reply by hertzsprung** [reply]
2.4.2 (I've edited my post, too)

**Reply by hertzsprung** [reply]
Aha, thank you!  Perhaps verging off topic, but should a `@SqsListener` method reasonably expect to handle a `null` payload?  Less surprising if Spring Cloud AWS threw an exception without invoking the listener method.  But perhaps this is already being fixed in 3.0? :slightly_smiling_face:

**Reply by maciejwalkowiak** [reply]
I think framework shouldn't try to be too smart about it. Payload can be null but headers can be there. I think I prefer to leave it to the user space.
